### PR TITLE
Update System.Net.Security Telemetry tests

### LINKED ISSUE #42613: Test failure: System.Net.Security.Tests.TelemetryTest.EventSource_SuccessfulHandshake_LogsStartStop
failed in job: [runtime-libraries-coreclr outerloop 20200922.1 ](https://dev.azure.com/dnceng/public/_build/results?buildId=825084&view=ms.vss-test-web.build-test-results-tab&runId=26282208&paneView=debug&resultId=102213)

net5.0-OSX-Release-x64-CoreCLR_release-OSX.1014.Amd64.Open

Error message
~~~
Microsoft.DotNet.RemoteExecutor.RemoteExecutionException : Remote process failed with an unhandled exception.


Stack trace

Child exception:
  Xunit.Sdk.ContainsException: Assert.Contains() Failure
Not found: (filter expression)
In value:  Double[] [0, 0, 0, 0]
   at System.Net.Security.Tests.TelemetryTest.VerifyEventCounters(ConcurrentQueue`1 events, Boolean shouldHaveFailures) in /_/src/libraries/System.Net.Security/tests/FunctionalTests/TelemetryTest.cs:line 144
   at System.Net.Security.Tests.TelemetryTest.<>c.<<EventSource_SuccessfulHandshake_LogsStartStop>b__1_0>d.MoveNext() in /_/src/libraries/System.Net.Security/tests/FunctionalTests/TelemetryTest.cs:line 59
--- End of stack trace from previous location ---
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 65

Child process:
  System.Net.Security.Tests, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 System.Net.Security.Tests.TelemetryTest+<>c System.Threading.Tasks.Task <EventSource_SuccessfulHandshake_LogsStartStop>b__1_0()
~~~

### PR DESCRIPTION
**Pull request by MihaZupan**
Updated the NetSecurity telemetry tests to use a `WaitForEventCounters` helper instead of relying on a racy `Task.Delay`.
Enhanced the test to also check for correct ActivityIds.

Updated the product instrumentation to use a real async method instead of `.ContinueWith` to ensure events fire under the right ExecutionContext. Similar to the fix in `NameResolution` from #42188 (https://github.com/dotnet/runtime/issues/41670#issuecomment-685038205).

Fixes #42613

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 1]
```diff
@@ -194,7 +194,17 @@ private SecurityStatusPal EncryptData(ReadOnlyMemory<byte> buffer, ref byte[] ou
...
+                ValueTask task = ProcessAuthenticationWithTelemetryAsync(isAsync, isApm, cancellationToken);
+                if (isAsync)
+                {
+                    return task.AsTask();
+                }
+                else
+                {
+                    Debug.Assert(task.IsCompleted);
+                    task.GetAwaiter().GetResult();
+                    return null;
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs:206]: I don't remember why we did this, but seeing it again now, it'd be cleaner/simpler to just always return the Task and then have the synchronous call sites do the .GetAwaiter().GetResult().

Comment (MihaZupan) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Changed to always return the Task, looks cleaner overall.

**Comment by karelz** [comment]
@MihaZupan is it ready for re-review from @stephentoub?

**Comment by MihaZupan** [comment]
Yes. I pushed the change to always return a `Task` from `ProcessAuthenticationAsync` as suggested, which should make the logic cleaner.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 2]
```diff
@@ -365,7 +365,7 @@ public Task AuthenticateAsClientAsync(SslClientAuthenticationOptions sslClientAu
 
             ValidateCreateContext(sslClientAuthenticationOptions, _userCertificateValidationCallback, _certSelectionDelegate);
 
-            return ProcessAuthentication(true, false, cancellationToken)!;
+            return ProcessAuthenticationAsync(true, false, cancellationToken);
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs:368]: Nit: as long as you're editing the code, can you name the Boolean arguments?
```suggestion
            return ProcessAuthenticationAsync(isAsync: true, isApm: false, cancellationToken);
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 3]
```diff
@@ -375,7 +375,7 @@ private Task AuthenticateAsClientApm(SslClientAuthenticationOptions sslClientAut
 
             ValidateCreateContext(sslClientAuthenticationOptions, _userCertificateValidationCallback, _certSelectionDelegate);
 
-            return ProcessAuthentication(true, true, cancellationToken)!;
+            return ProcessAuthenticationAsync(true, true, cancellationToken);
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs:378]:
```suggestion
            return ProcessAuthenticationAsync(isAsync: true, isApm: true, cancellationToken);
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 4]
```diff
@@ -418,21 +418,21 @@ public Task AuthenticateAsServerAsync(SslServerAuthenticationOptions sslServerAu
             SetAndVerifyValidationCallback(sslServerAuthenticationOptions.RemoteCertificateValidationCallback);
             ValidateCreateContext(CreateAuthenticationOptions(sslServerAuthenticationOptions));
 
-            return ProcessAuthentication(true, false, cancellationToken)!;
+            return ProcessAuthenticationAsync(true, false, cancellationToken);
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs:421]:
```suggestion
            return ProcessAuthenticationAsync(isAsync: true, isApm: false, cancellationToken);
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 5]
```diff
@@ -418,21 +418,21 @@ public Task AuthenticateAsServerAsync(SslServerAuthenticationOptions sslServerAu
...
         }
 
         private Task AuthenticateAsServerApm(SslServerAuthenticationOptions sslServerAuthenticationOptions, CancellationToken cancellationToken = default)
         {
             SetAndVerifyValidationCallback(sslServerAuthenticationOptions.RemoteCertificateValidationCallback);
             ValidateCreateContext(CreateAuthenticationOptions(sslServerAuthenticationOptions));
 
-            return ProcessAuthentication(true, true, cancellationToken)!;
+            return ProcessAuthenticationAsync(true, true, cancellationToken);
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs:429]:
```suggestion
            return ProcessAuthenticationAsync(isAsync: true, isApm: true, cancellationToken);
```

**Review by stephentoub — APPROVED** [review]
Thanks.

**Pull request merged**